tests/test_0652-tests-of-complex-numbers.py

```diff
@@ -0,0 +1,214 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+from __future__ import absolute_import
+
+import json
+
+import pytest  # noqa: F401
+import numpy as np  # noqa: F401
+import awkward as ak  # noqa: F401
+
+
+def test_from_iter():
+    assert ak.from_iter([1 + 1j, 2 + 2j, 3 + 3j]).tolist() == [1 + 1j, 2 + 2j, 3 + 3j]
+    assert ak.from_iter([[1 + 1j, 2 + 2j], [], [3 + 3j]]).tolist() == [
+        [1 + 1j, 2 + 2j],
+        [],
+        [3 + 3j],
+    ]
+
+    # First encounter of a complex number should promote previous integers and
+    # reals into complex numbers:
+    assert ak.from_iter([1, 2.2, 3 + 3j]).tolist() == [1.0 + 0j, 2.2 + 0j, 3.0 + 3j]
+    assert ak.from_iter([1, 3 + 3j]).tolist() == [1.0 + 0j, 3.0 + 3j]
+
+    # Just as the first encounter of a real number promotes previous integers
+    # into reals:
+    assert str(ak.from_iter([1, 2.2]).type) == "2 * float64"
+    assert ak.from_iter([1, 2.2]).tolist() == [1.0, 2.2]
+    builder = ak.ArrayBuilder()
+    str(ak.type(builder)) == "0 * unknown"
+    builder.integer(1)
+    str(ak.type(builder)) == "1 * int64"
+    builder.real(2.2)
+    str(ak.type(builder)) == "2 * float64"
+
+    # For that matter, ArrayBuilder is missing a high-level interface to complex:
+    builder.complex(3 + 3j)
+    str(ak.type(builder)) == "3 * complex128"
+
+
+def test_from_json():
+    array = ak.from_json('[{"r": 1.1, "i": 1.0}, {"r": 2.2, "i": 2.0}]')
+    assert array.tolist() == [
+        {"r": 1.1, "i": 1.0},
+        {"r": 2.2, "i": 2.0},
+    ]
+    array = ak.from_json(
+        '[{"r": 1.1, "i": 1.0}, {"r": 2.2, "i": 2.0}]', complex_record_fields=("r", "i")
+    )
+    assert (array["r"] + array["i"]*1j).tolist() == [(1.1 + 1j), (2.2 + 2j)]
```

Comment:
The logic that performs this substitution should be inside the `ak.from_json` Python function, though not inside the `Content::fromjson` C++ function. The test should be reverted to

```python
    array = ak.from_json(
￼        '[{"r": 1.1, "i": 1.0}, {"r": 2.2, "i": 2.0}]', complex_record_fields=("r", "i")
￼    )
￼    assert array.tolist() == [(1.1 + 1j), (2.2 + 2j)]
```

The `ak.from_json` function can be made to do this using the trick you described. To get it to work at any level, such as in this test:

```python
    array = ak.from_json(
￼        '[[{"r": 1.1, "i": 1.0}, {"r": 2.2, "i": 2.0}], [], [{"r": 3.3, "i": 3.0}]]', complex_record_fields=("r", "i")
￼    )
￼    assert array.tolist() == [[(1.1 + 1j), (2.2 + 2j)], [], [(3.3 + 3j)]]
```

by using the internal utility function [ak._util.recursively_apply](https://github.com/scikit-hep/awkward-1.0/blob/main/src/awkward/_util.py#L1068-L1076). Here is a sample usage:

https://github.com/scikit-hep/awkward-1.0/blob/3908bff17f4552b4dad3299cfcee66a1aef77735/src/awkward/operations/structure.py#L574-L590

In the above, the `getfunction` only returns non-None when it encounters a RecordArray at some depth. Yours would return non-None when it encounters a RecordArray with at least the fields specified by `complex_record_fields`. The above returns a zero-argument lambda that returns a modified RecordArray. Yours would return a NumpyArray of complex type.

There's an open question of what to do if you find a record with fields specified by `complex_record_fields` that are not one-dimensional NumpyArrays. Should the substitution just not apply or should it raise an error? I'm inclined to think it should just not apply.

Also, although the

```python
return lambda: recordnode[complex_record_fields[0]] + recordnode[complex_record_fields[1]]*1j
```

is slick, the following crosses all the "t"s and dots all the "i"s for alternate backends:

```python
real = recordnode[complex_record_fields[0]]
imag = recordnode[complex_record_fields[1]]
nplike = ak.nplike.of(real, imag)
real_array = nplike.asarray(real)
imag_array = nplike.asarray(imag)
float_dtype = ak.nplike.numpy.result_type(real_array.dtype, imag_array.dtype, np.dtype(np.float32))
out = nplike.empty(2*len(recordnode), dtype=float_dtype)
out[::2] = real_array
out[1::2] = imag_array
complex_dtype = {
    np.dtype(np.float32): np.complex64,
    np.dtype(np.float64): np.complex128
}[float_dtype]
return lambda: ak.layout.NumpyArray(out.view(complex_dtype))
```

I've been thinking about it and the slick solution is probably just as good as this long one. I was worried that the slick solution would over-promote the complex type, returning a `np.complex128` when a `np.complex64` is desired, but the ArrayBuilder always returns the largest floating-point type, so this isn't an issue.

I guess you should stick with the slick solution: multiplication and addition. It's certainly easier to read.

Reply:
Oh, I'm missing something needed for multiplication:
```python
TypeError: unsupported operand type(s) for *: 'awkward._ext.NumpyArray' and 'complex'
```

Reply:
All of the operator overloads like `__mul__` are implemented on `ak.Array`, not the layout node classes. But this could be a better way to do it than re-wrapping those NumpyArrays as `ak.Arrays`:

```python
nplike = ak.nplike.of(recordnode)
real = nplike.asarray(recordnode[complex_record_fields[0]])
imag = nplike.asarray(recordnode[complex_record_fields[1]])
return lambda: ak.layout.NumpyArray(real + imag*1j)
```

This avoids a lot of unnecessary infrastructure that `ak.Array` has to go through to implement `__add__` and `__mul__` on arbitrary structures. Since we know that these are both 1-dimensional NumPy arrays with equal lengths, casting them as such is more direct. (Although they might be GPU arrays to be handled by CuPy, but that's why we use `nplike`.)

Reply:
@jpivarski - thanks! It looks nicer. Done and dusted :-)